TODO: SparkBatchSink/SparkBatchSource factory logic also needs to change to account for same aliases. Here is the error I see when I try to execute a spark pipeline with sinks (with same reference names):
```
co.cask.cdap.app.runtime.spark.SparkRuntimeService$3$1.run(SparkRuntimeService.java:300) [co.cask.cdap.cdap-spark-core-3.5.0-SNAPSHOT.jar:na] at java.lang.Thread.run(Thread.java:745) [na:1.7.0_75] Caused by: java.lang.IllegalArgumentException: Output already configured: bighdfs at co.cask.cdap.etl.batch.spark.SparkBatchSinkFactory.addOutput(SparkBatchSinkFactory.java:122) ~[na:na] at co.cask.cdap.etl.batch.spark.SparkBatchSinkFactory.addOutput(SparkBatchSinkFactory.java:90) ~[na:na] at co.cask.cdap.etl.batch.spark.SparkBatchSinkContext.addOutput(SparkBatchSinkContext.java:60) ~[na:na] at co.cask.hydrator.plugin.HDFSSink.prepareRun(HDFSSink.java:73) ~[na:na] at co.cask.hydrator.plugin.HDFSSink.prepareRun(HDFSSink.java:50) ~[na:na] at co.cask.cdap.etl.batch.spark.ETLSpark.beforeSubmit(ETLSpark.java:133) ~[na:na] at co.cask.cdap.app.runtime.spark.SparkRuntimeService$4.call(SparkRuntimeService.java:321) ~[co.cask.cdap.cdap-spark-core-3.5.0-SNAPSHOT.jar:na] at co.cask.cdap.app.runtime.spark.SparkRuntimeService$4.call(SparkRuntimeService.java:316) ~[co.cask.cdap.cdap-spark-core-3.5.0-SNAPSHOT.jar:na] at co.cask.cdap.data2.transaction.Transactions.execute(Transactions.java:174) ~[co.cask.cdap.cdap-data-fabric-3.5.0-SNAPSHOT.jar:na] ... 5 common frames omitted
```

--- a/cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/ETLSpark.java
+++ b/cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/ETLSpark.java
@@ -108,9 +108,7 @@ public class ETLSpark extends AbstractSpark {
 
     BatchConfigurable<BatchSourceContext> batchSource = pluginInstantiator.newPluginInstance(sourceName);
     DatasetContextLookupProvider lookProvider = new DatasetContextLookupProvider(context);
-    SparkBatchSourceContext sourceContext = new SparkBatchSourceContext(context,
-                                                                        lookProvider,
-                                                                        sourceName);
+    SparkBatchSourceContext sourceContext = new SparkBatchSourceContext(context, lookProvider, sourceName);
     batchSource.prepareRun(sourceContext);
 
     SparkBatchSourceFactory sourceFactory = sourceContext.getSourceFactory();

--- a/cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSinkContext.java
+++ b/cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSinkContext.java
@@ -25,6 +25,7 @@ import co.cask.cdap.etl.common.ExternalDatasets;
 
 import java.util.Collections;
 import java.util.Map;
+import java.util.UUID;
 
 /**
  * Default implementation of {@link BatchSinkContext} for spark contexts.
@@ -46,17 +47,25 @@ public class SparkBatchSinkContext extends AbstractSparkBatchContext implements 
 
   @Override
   public void addOutput(String datasetName, Map<String, String> arguments) {
-    sinkFactory.addOutput(getStageName(), datasetName, arguments);
+    sinkFactory.addOutput(getStageName(), suffixOutput(Output.ofDataset(datasetName, arguments)));
   }
 
   @Override
   public void addOutput(String outputName, OutputFormatProvider outputFormatProvider) {
-    sinkFactory.addOutput(getStageName(), outputName, outputFormatProvider);
+    sinkFactory.addOutput(getStageName(), suffixOutput(Output.of(outputName, outputFormatProvider)));
   }
 
   @Override
   public void addOutput(Output output) {
-    Output trackableOutput = ExternalDatasets.makeTrackable(sparkContext.getAdmin(), output);
+    Output trackableOutput = ExternalDatasets.makeTrackable(sparkContext.getAdmin(), suffixOutput(output));
     sinkFactory.addOutput(getStageName(), trackableOutput);
+  }
+
+  /**
+   * Suffix the alias of {@link Output} so that aliases of outputs are unique.
+   */
+  private Output suffixOutput(Output output) {
+    String suffixedAlias = String.format("%s-%s", output.getAlias(), UUID.randomUUID());
+    return output.alias(suffixedAlias);
   }
 }

--- a/cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSinkFactory.java
+++ b/cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSinkFactory.java
@@ -98,16 +98,6 @@ final class SparkBatchSinkFactory {
     }
   }
 
-  void addOutput(String stageName, String alias, OutputFormatProvider outputFormatProvider) {
-    addOutput(stageName, alias,
-              new BasicOutputFormatProvider(outputFormatProvider.getOutputFormatClassName(),
-                                            outputFormatProvider.getOutputFormatConfiguration()));
-  }
-
-  void addOutput(String stageName, String datasetName, Map<String, String> datasetArgs) {
-    addOutput(stageName, datasetName, datasetName, datasetArgs);
-  }
-
   private void addOutput(String stageName, String alias,
                          BasicOutputFormatProvider outputFormatProvider) {
     if (outputFormatProviders.containsKey(alias) || datasetInfos.containsKey(alias)) {

--- a/cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSourceContext.java
+++ b/cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSourceContext.java
@@ -16,12 +16,10 @@
 
 package co.cask.cdap.etl.batch.spark;
 
-import co.cask.cdap.api.data.batch.BatchReadable;
 import co.cask.cdap.api.data.batch.Input;
 import co.cask.cdap.api.data.batch.InputFormatProvider;
 import co.cask.cdap.api.data.batch.Split;
 import co.cask.cdap.api.data.stream.StreamBatchReadable;
-import co.cask.cdap.api.dataset.Dataset;
 import co.cask.cdap.api.spark.SparkClientContext;
 import co.cask.cdap.etl.api.LookupProvider;
 import co.cask.cdap.etl.api.batch.BatchSourceContext;
@@ -36,7 +34,6 @@ import javax.annotation.Nullable;
  * Default implementation of {@link BatchSourceContext} for spark contexts.
  */
 public class SparkBatchSourceContext extends AbstractSparkBatchContext implements BatchSourceContext {
-
   private SparkBatchSourceFactory sourceFactory;
 
   public SparkBatchSourceContext(SparkClientContext sparkContext, LookupProvider lookupProvider, String stageId) {
@@ -50,17 +47,17 @@ public class SparkBatchSourceContext extends AbstractSparkBatchContext implement
 
   @Override
   public void setInput(String datasetName) {
-    sourceFactory = SparkBatchSourceFactory.create(datasetName);
+    setInput(datasetName, Collections.<String, String>emptyMap());
   }
 
   @Override
   public void setInput(String datasetName, Map<String, String> arguments) {
-    sourceFactory = SparkBatchSourceFactory.create(datasetName, arguments);
+    setInput(datasetName, arguments, null);
   }
 
   @Override
   public void setInput(String datasetName, List<Split> splits) {
-    sourceFactory = SparkBatchSourceFactory.create(datasetName, Collections.<String, String>emptyMap(), splits);
+    setInput(datasetName, Collections.<String, String>emptyMap(), splits);
   }
 
   @Override

--- a/cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSourceFactory.java
+++ b/cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSourceFactory.java
@@ -83,14 +83,6 @@ final class SparkBatchSourceFactory {
     return new SparkBatchSourceFactory(null, inputFormatProvider, null);
   }
 
-  static SparkBatchSourceFactory create(String datasetName) {
-    return create(datasetName, ImmutableMap.<String, String>of());
-  }
-
-  static SparkBatchSourceFactory create(String datasetName, Map<String, String> datasetArgs) {
-    return create(datasetName, datasetArgs, null);
-  }
-
   static SparkBatchSourceFactory create(String datasetName, Map<String, String> datasetArgs,
                                         @Nullable List<Split> splits) {
     return new SparkBatchSourceFactory(null, null, new DatasetInfo(datasetName, datasetArgs, splits));

--- a/cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceSinkContext.java
+++ b/cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceSinkContext.java
@@ -25,9 +25,11 @@ import co.cask.cdap.etl.api.batch.BatchSinkContext;
 import co.cask.cdap.etl.common.ExternalDatasets;
 import co.cask.cdap.etl.log.LogContext;
 
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
+import java.util.UUID;
 import java.util.concurrent.Callable;
 
 /**
@@ -45,38 +47,33 @@ public class MapReduceSinkContext extends MapReduceBatchContext implements Batch
 
   @Override
   public void addOutput(final String datasetName) {
-    LogContext.runWithoutLoggingUnchecked(new Callable<Void>() {
-      @Override
-      public Void call() throws Exception {
-        mrContext.addOutput(datasetName);
-        return null;
-      }
-    });
-    outputNames.add(datasetName);
+    addOutput(datasetName, Collections.<String, String>emptyMap());
   }
 
   @Override
   public void addOutput(final String datasetName, final Map<String, String> arguments) {
-    LogContext.runWithoutLoggingUnchecked(new Callable<Void>() {
+    String alias = LogContext.runWithoutLoggingUnchecked(new Callable<String>() {
       @Override
-      public Void call() throws Exception {
-        mrContext.addOutput(datasetName, arguments);
-        return null;
+      public String call() throws Exception {
+        Output output = suffixOutput(Output.ofDataset(datasetName, arguments));
+        mrContext.addOutput(output);
+        return output.getAlias();
       }
     });
-    outputNames.add(datasetName);
+    outputNames.add(alias);
   }
 
   @Override
   public void addOutput(final String outputName, final OutputFormatProvider outputFormatProvider) {
-    LogContext.runWithoutLoggingUnchecked(new Callable<Void>() {
+    String alias = LogContext.runWithoutLoggingUnchecked(new Callable<String>() {
       @Override
-      public Void call() throws Exception {
-        mrContext.addOutput(outputName, outputFormatProvider);
-        return null;
+      public String call() throws Exception {
+        Output output = suffixOutput(Output.of(outputName, outputFormatProvider));
+        mrContext.addOutput(output);
+        return output.getAlias();
       }
     });
-    outputNames.add(outputName);
+    outputNames.add(alias);
   }
 
   @Override
@@ -84,7 +81,7 @@ public class MapReduceSinkContext extends MapReduceBatchContext implements Batch
     Output trackableOutput = LogContext.runWithoutLoggingUnchecked(new Callable<Output>() {
       @Override
       public Output call() throws Exception {
-        Output trackableOutput = ExternalDatasets.makeTrackable(mrContext.getAdmin(), output);
+        Output trackableOutput = ExternalDatasets.makeTrackable(mrContext.getAdmin(), suffixOutput(output));
         mrContext.addOutput(trackableOutput);
         return trackableOutput;
       }
@@ -97,5 +94,13 @@ public class MapReduceSinkContext extends MapReduceBatchContext implements Batch
    */
   public Set<String> getOutputNames() {
     return outputNames;
+  }
+
+  /**
+   * Suffix the alias of {@link Output} so that aliases of outputs are unique.
+   */
+  private Output suffixOutput(Output output) {
+    String suffixedAlias = String.format("%s-%s", output.getAlias(), UUID.randomUUID());
+    return output.alias(suffixedAlias);
   }
 }

--- a/cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceSourceContext.java
+++ b/cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceSourceContext.java
@@ -28,12 +28,13 @@ import co.cask.cdap.etl.api.batch.BatchSourceContext;
 import co.cask.cdap.etl.common.ExternalDatasets;
 import co.cask.cdap.etl.log.LogContext;
 
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.Callable;
 
 /**
- * MapReduce Source Context.
+ * MapReduce Source Context. Delegates operations to MapReduce Context.
  */
 public class MapReduceSourceContext extends MapReduceBatchContext implements BatchSourceContext {
 
@@ -61,35 +62,17 @@ public class MapReduceSourceContext extends MapReduceBatchContext implements Bat
 
   @Override
   public void setInput(final String datasetName) {
-    LogContext.runWithoutLoggingUnchecked(new Callable<Void>() {
-      @Override
-      public Void call() throws Exception {
-        mrContext.addInput(Input.ofDataset(datasetName));
-        return null;
-      }
-    });
+    setInput(datasetName, Collections.<String, String>emptyMap());
   }
 
   @Override
   public void setInput(final String datasetName, final Map<String, String> arguments) {
-    LogContext.runWithoutLoggingUnchecked(new Callable<Void>() {
-      @Override
-      public Void call() throws Exception {
-        mrContext.addInput(Input.ofDataset(datasetName, arguments));
-        return null;
-      }
-    });
+    setInput(datasetName, arguments, null);
   }
 
   @Override
   public void setInput(final String datasetName, final List<Split> splits) {
-    LogContext.runWithoutLoggingUnchecked(new Callable<Void>() {
-      @Override
-      public Void call() throws Exception {
-        mrContext.addInput(Input.ofDataset(datasetName, splits));
-        return null;
-      }
-    });
+    setInput(datasetName, Collections.<String, String>emptyMap(), splits);
   }
 
   @Override

